syntax = "proto3";
option java_multiple_files = true;
option java_outer_classname = "SeServiceProto";

package wood.mike;

service SecretEscapes {
  // unary
  rpc GetSeSale (SeSaleRequest) returns (SeSaleReply);

  // server streaming
  rpc GetAllSeSales(SeSaleRangeRequest) returns (stream SeSaleReply);

  // client streaming
  rpc GetBatchedSeSales(stream SeSaleRequest) returns (BatchSeSaleReply);

  // bidirectional streaming
  rpc CustomizePackage(stream PackageRequest) returns (stream PackageResponse);
}

message SeSaleRequest {
  int32 id = 1;
}

message SeSaleRangeRequest {
  string start = 1;
  string end = 2;
}

message SeSaleReply {
  int32 id = 1;
  string url_slug = 2;
  string start = 3;
  string end = 4;
}

message BatchSeSaleReply {
  repeated SeSaleReply replies = 1;
}

message PackageRequest {
  string user_id = 1;
  string selection_type = 2;  // e.g., "flight", "hotel", "tour"
  string selection_id = 3;    // ID of the selected option
  int32 quantity = 4;         // For things like number of nights, tickets, etc.
}

// Response message with availability and price feedback
message PackageResponse {
  string selection_type = 1;
  string selection_id = 2;
  bool available = 3;
  double price = 4;
  string message = 5;  // e.g., "Available", "Sold out", "Price updated"
}